samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java

```diff
@@ -48,141 +49,184 @@
+          .collect(Collectors.toList()));
+    } else {
+      String msg = "Avro Record is null";
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    return new SamzaSqlRelRecord(fieldNames, values);
   }
 
+  /**
+   * Convert the nested relational message to the output samza message.
+   */
   @Override
   public KV<Object, Object> convertToSamzaMessage(SamzaSqlRelMessage relMessage) {
     return convertToSamzaMessage(relMessage, this.avroSchema);
   }
 
   protected KV<Object, Object> convertToSamzaMessage(SamzaSqlRelMessage relMessage, Schema avroSchema) {
-    GenericRecord record = new GenericData.Record(avroSchema);
-    List<String> fieldNames = relMessage.getFieldNames();
-    List<Object> values = relMessage.getFieldValues();
+    return new KV<>(relMessage.getKey(), convertToGenericRecord(relMessage.getSamzaSqlRelRecord(), avroSchema));
+  }
+
+  private GenericRecord convertToGenericRecord(SamzaSqlRelRecord relRecord, Schema schema) {
+    GenericRecord record = new GenericData.Record(schema);
+    List<String> fieldNames = relRecord.getFieldNames();
+    List<Object> values = relRecord.getFieldValues();
     for (int index = 0; index < fieldNames.size(); index++) {
       if (!fieldNames.get(index).equalsIgnoreCase(SamzaSqlRelMessage.KEY_NAME)) {
-        record.put(fieldNames.get(index), values.get(index));
+        Object relObj = values.get(index);
+        String fieldName = fieldNames.get(index);
+        Schema fieldSchema = schema.getField(fieldName).schema();
+        record.put(fieldName, convertToAvroObject(relObj, fieldSchema));
       }
     }
+    return record;
+  }
 
-    return new KV<>(relMessage.getKey(), record);
+  private Object convertToAvroObject(Object relObj, Schema schema) {
+    if (relObj == null) {
+      return null;
+    }
+    switch(schema.getType()) {
+      case RECORD:
+        return convertToGenericRecord((SamzaSqlRelRecord) relObj, getNonNullUnionSchema(schema));
+      case ARRAY:
+        List<Object> avroList = ((List<Object>) relObj).stream()
+            .map(o -> convertToAvroObject(o, getNonNullUnionSchema(schema).getElementType()))
+            .collect(Collectors.toList());
+        return avroList;
+      case MAP:
+        return ((Map<String, ?>) relObj).entrySet()
+            .stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, e -> convertToAvroObject(e.getValue(),
+                getNonNullUnionSchema(schema).getValueType())));
+      case UNION:
+        return convertToAvroObject(relObj, getNonNullUnionSchema(schema));
+      default:
+        return relObj;
+    }
   }
 
-  private Object getRelField(RelDataType relType, Object avroObj) {
-    return AvroToRelObjConverter.valueOf(relType.getClass().getSimpleName()).convert(avroObj);
+  private Object convertToJavaObject(Object avroObj, Schema schema) {
+    switch(schema.getType()) {
+      case RECORD:
+        if (avroObj == null) {
+          return null;
```

Comment:
Can you open a bug and also TODO to perform validations of the data types with the Avro schema here?

Reply:
As discussed offline, I have added a comment saying that we are not adding validations here considering the cost per message.